engine/access/rest/websockets/controller.go

```diff
@@ -202,20 +228,21 @@ func (c *Controller) handleListSubscriptions(ctx context.Context, msg models.Lis
 }
 
 func (c *Controller) shutdownConnection() {
-	defer close(c.communicationChannel)
-	defer func(conn *websocket.Conn) {
-		if err := c.conn.Close(); err != nil {
-			c.logger.Error().Err(err).Msg("error closing connection")
-		}
-	}(c.conn)
+	c.shutdownOnce.Do(func() {
+		defer func() {
+			if err := c.conn.Close(); err != nil {
+				c.logger.Warn().Err(err).Msg("error closing connection")
+			}
+		}()
 
-	err := c.dataProviders.ForEach(func(_ uuid.UUID, dp dp.DataProvider) error {
-		dp.Close()
-		return nil
-	})
-	if err != nil {
-		c.logger.Error().Err(err).Msg("error closing data provider")
-	}
+		c.logger.Debug().Msg("shutting down connection")
 
-	c.dataProviders.Clear()
+		_ = c.dataProviders.ForEach(func(id uuid.UUID, dp dp.DataProvider) error {
+			err := dp.Close()
+			c.logger.Error().Err(err).Msgf("error closing data provider: %s", id.String())
```

Comment:
use labels where possible
```suggestion
			c.logger.Error().Err(err).
				Str("data_provider", id.String()).
				Msg("error closing data provider")
```